### PR TITLE
Add class constraints for built-in classes

### PR DESCRIPTION
**Pull request by sezna**
This PR exposes some relevant built-in classes via class constraints.

# Changes
- Add `TypeParameter`, `ClassConstraint`, and `ConstraintParameter` AST nodes to represent the notion of a generic type that might have class constraints, and constraints that may have parameters (e.g. `Exp[Int]`).
  -  update visitors accordingly
- `Ty` can now contain arbitrary constraints
- Add notion of primitive classes to the AST
- Update constraint checking algorithm to check type parameters for primitive constraints at callable decl time
- Update constraint generation algorithm to add additional user-specified constraints to arguments that are passed in to callables (`fn constrained_ty`)
- Add completions for primitive classes to the language service
- Adds a sample describing class constraints

tiny changes:
- Add docstrings discretionally
- Remove some unnecessary derives
- Tried to clarify and unify jargon around classes, constraints, parameters, and arguments in general
- Some expect_tests were updated in minor ways as their underlying types changed (`Ty` and `TyParam`, for example)
- Improve an error message for `MissingTy` (it mentioned something about global types not being inferred when 99% of the time it does not apply to a global type, it applies to a callable parameter with a missing type)

### CODE REVIEW

File: compiler/qsc_frontend/src/typeck/convert.rs [hunk 1]
```diff
@@ -189,18 +227,48 @@ pub(super) fn ast_ty_def(names: &Names, def: &TyDef) -> (UdtDef, Vec<MissingTyEr
...
+    names: &Names,
+    generics: &[ast::TypeParameter],
+) -> (Vec<qsc_hir::ty::GenericParam>, Vec<TyConversionError>) {
+    let mut errors = Vec::new();
+    let mut generics_buf = Vec::with_capacity(generics.len());
+    for param in generics {
+        let (bounds, new_errors) =
+            class_constraints_from_ast(names, &param.constraints, &mut Default::default());
+        errors.extend(new_errors);
+        generics_buf.push(GenericParam::Ty {
```
Comment (sezna) [inline compiler/qsc_frontend/src/typeck/convert.rs:242]: @sezna rename these

File: compiler/qsc_frontend/src/typeck/infer.rs [hunk 2]
```diff
@@ -383,6 +411,20 @@ impl Inferrer {
         self.constraints.push_back(Constraint::Class(class, span));
     }
 
+    /// Returns a unique type variable with specified constraints.
+    fn constrained_ty(
+        &mut self,
+        meta: TySource,
+        with_constraints: impl Fn(Ty) -> Vec<Constraint>,
```
Comment (sezna) [inline compiler/qsc_frontend/src/typeck/infer.rs:418]: consider `.into_boxed_slice()`

File: compiler/qsc_frontend/src/typeck/infer.rs [hunk 3]
```diff
@@ -891,38 +998,80 @@ fn check_eq(ty: Ty, span: Span) -> (Vec<Constraint>, Vec<Error>) {
...
+        }
         _ => (
             Vec::new(),
             vec![Error(ErrorKind::MissingClassExp(base.display(), span))],
         ),
     }
 }
 
+// i'm using the wildcard below to enforce that Ty::Param is always matched in the err branch, as
+// it shouldn't be constrained by HasField as long as we don't support structural typing
```
Comment (sezna) [inline compiler/qsc_frontend/src/typeck/infer.rs:1073]: @sezna

File: compiler/qsc_lowerer/src/lib.rs [hunk 4]
```diff
@@ -878,10 +878,57 @@ impl Lowerer {
             name_span: field.name_span,
         }
     }
-}
 
-fn lower_generics(generics: &[qsc_hir::ty::GenericParam]) -> Vec<qsc_fir::ty::GenericParam> {
-    generics.iter().map(lower_generic_param).collect()
+    fn lower_generics(
```
Comment (sezna) [inline compiler/qsc_lowerer/src/lib.rs:882]: This had to become a method, because a type parameter can now contain class constraint parameters which could be concrete types. Those concrete types need to be lowered. `.lower_ty()` is a method, so we therefore have to make this a method to have access to `self`.

File: compiler/qsc_ast/src/ast.rs [hunk 5]
```diff
@@ -1431,6 +1435,28 @@ impl Default for PathKind {
...
+    #[must_use]
+    pub fn name(&self) -> Option<&Ident> {
+        match self {
+            PathKind::Ok(path) => Some(&path.name),
+            PathKind::Err(Some(incomplete_path)) => incomplete_path.segments.last(),
+            PathKind::Err(None) => None,
+        }
+    }
+}
+
```
Comment (minestarks) [inline compiler/qsc_ast/src/ast.rs:1459]: I had reservations about adding these methods back when I first defined `PathKind` because I couldn't think of a good reason you'd want to access the segments or name part of the path, without knowing if it's an `Err` or `Ok` first. Besides, there's always the `Idents` impl if you want to straightforwardly iterate over the contained `Ident`s.

On a more practical note, these are unused anyway, so maybe just delete?

File: compiler/qsc_ast/src/ast.rs [hunk 6]
```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
...
+/// A list of class constraints, used when constraining a type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash)]
+pub struct ClassConstraints(pub Box<[ClassConstraint]>);
+
+/// An individual class constraint, used when constraining a type parameter.
+/// To understand this concept, think of parameters in a function signature -- the potential arguments that can
+/// be passed to them are constrained by what type is specified. Type-level parameters are no different, and
+/// the type variables that are passed to a type parameter must satisfy the constraints specified in the type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash, Debug)]
```
Comment (minestarks) [inline compiler/qsc_ast/src/ast.rs:2057]: I've been trying to squash `Default` impls for the AST nodes because I think they also lead to bugs. I haven't been successful in eliminating all of them, but it'd be nice if we can stop adding more. It's too easy to end up with `""` idents in the AST, and then have the resolver spew nonsense errors like "`""` is not a name` etc. I think the compiler code would just be easier to work with if we stuck to invariants such as "no empty string identifiers". 

This one seems to be unused, so I think you can just delete it.

File: language_service/src/completion/tests.rs
Comment (minestarks) [inline language_service/src/completion/tests.rs:1]: Commenting in this random file, but this is about the language service in general:

Do you want to update the user-friendly function signature display to include the class bounds?

image

Seems like the hover text here would be better to show `function F<'x: Exp[Int] + Num>(x: 'x) : Unit`

File: samples/language/ClassConstraints.qs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+function Main() : Unit {
+    Message($"{AllEqual([1, 1, 1])}");
+    Message($"{AllEqual([1, 2, 3])}");
+
+    // Because we wrote this function generically, we are able to pass in different types, as
+    // long as they can be compared via the class `Eq`.
+    Message($"{AllEqual([true, true, false])}");
+    Message($"{AllEqual(["a", "b"])}");
```
Comment (swernli) [inline samples/language/ClassConstraints.qs:40]: nitpick: some of these are hard to parse because of the way string interpolation and syntax highlighting play together, so I think it would be clearer to pull the function invocations up out of the string, bind the result and just print that.

File: language_service/src/name_locator.rs
Comment (minestarks) [inline language_service/src/name_locator.rs:1]: Not sure where exactly the bug is, but there's a bug in the inline rename feature that's likely originating from this file somewhere.

```qsharp
function F<'x: Exp[Int] + Num>(x: 'x) : Unit {
}
```

Place the cursor over `'x` and invoke the "rename symbol" command . Give it a new name. Note how the class constraints disappear . This is likely due to the rename span returned to the editor being too large (it should just contain the type variable but it contains the bounds too)

File: compiler/qsc_ast/src/ast.rs [hunk 8]
```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
...
+            self.ty,
+            if self.constraints.0.is_empty() {
+                ""
+            } else {
+                ": "
+            },
+            self.constraints
+        )
+    }
+}
```
Comment (minestarks) [inline compiler/qsc_ast/src/ast.rs:2047]: This and the `Debug` impl for `ClassConstraint` are the first custom `Debug` impls in this file. The norm is to define `Display` for AST nodes, and use that in tests / AST dumps . We can debate whether that's the "right" way but it's a pretty entrenched pattern. Can you just make these `Display` impls maybe?

File: compiler/qsc_parse/src/ty.rs [hunk 9]
```diff
@@ -72,9 +73,53 @@ pub(super) fn array_or_arrow(s: &mut ParserContext<'_>, mut lhs: Ty, lo: u32) ->
...
+/// `T: Eq + Iterator[Bool] + Class3`
+///     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ bounds
+fn class_constraints(s: &mut ParserContext) -> Result<ClassConstraints> {
+    let mut bounds: Vec<ClassConstraint> = Vec::new();
+    loop {
+        s.expect(WordKinds::PrimitiveClass);
+        let bound_name = ident(s)?;
+        // if there's a less-than sign, or "open angle bracket", try to parse type parameters for
+        // the class
+        // e.g. `Iterator[Bool]`
```
Comment (swernli) [inline compiler/qsc_parse/src/ty.rs:104]: code is correct, but comment is out of date, should be about angle brackets:

```suggestion
        // if there's an "open bracket", try to parse type parameters for the class
        // e.g. `Iterator[Bool]`
```

File: compiler/qsc_ast/src/ast.rs [hunk 10]
```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
...
+impl std::fmt::Debug for ClassConstraints {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // A + B + C + D
+        write!(
+            f,
+            "{}",
+            self.0
+                .iter()
+                .map(|x| x.name.to_string())
```
Comment (minestarks) [inline compiler/qsc_ast/src/ast.rs:2117]: This misses the constraint parameters in the debug dump.

File: compiler/qsc_ast/src/ast.rs [hunk 11]
```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
...
+        }
+    }
+}
+
+impl ConstraintParameter {
+    /// Getter for the `ty` field.
+    #[must_use]
+    pub fn ty(&self) -> &Ty {
+        &self.ty
+    }
```
Comment (minestarks) [inline compiler/qsc_ast/src/ast.rs:2088]: Why these trivial getters for public fields? Again I'd argue for consistency across the AST nodes - the existing convention seems to be accessing fields directly.

File: compiler/qsc_doc_gen/src/display.rs [hunk 12]
```diff
@@ -615,12 +615,12 @@ where
...
 }
 
-fn display_type_params(generics: &[GenericParam]) -> String {
+fn display_type_params(generics: &[HirTypeParameter]) -> String {
     let type_params = generics
         .iter()
         .filter_map(|generic| match generic {
-            GenericParam::Ty(name) => Some(name.name.clone()),
-            GenericParam::Functor(_) => None,
+            HirTypeParameter::Ty { name, .. } => Some(name.clone()),
```
Comment (swernli) [inline compiler/qsc_doc_gen/src/display.rs:622]: I think it's worth filing a follow up to make sure docs generation includes type bounds. Right now it's not critical since the stdlib doesn't have anything with bounds, but the generated docs for your new library (other PR) will be wrong in VS Code until this is updated.

File: compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs [hunk 13]
```diff
@@ -0,0 +1,822 @@
...
+            #42 167-172 "a: 'E" : Param<"'E": 0>
+            #46 174-180 "b: Int" : Int
+            #53 187-224 "{\n                a ^ b\n            }" : Param<"'E": 0>
+            #55 205-210 "a ^ b" : Param<"'E": 0>
+            #56 205-206 "a" : Param<"'E": 0>
+            #59 209-210 "b" : Int
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 2, span: Span { lo: 56, hi: 59 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 2, span: Span { lo: 56, hi: 59 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 0, span: Span { lo: 162, hi: 165 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 0, span: Span { lo: 162, hi: 165 } })))
```
Comment (swernli) [inline compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs:109]: Wasn't this the duplicate error bug you said you fixed? Or was there more than one and one of them was benign?

File: compiler/qsc_frontend/src/typeck.rs [hunk 14]
```diff
@@ -119,4 +124,60 @@ enum ErrorKind {
     #[diagnostic(help("provide a type annotation"))]
     #[diagnostic(code("Qsc.TypeCk.AmbiguousTy"))]
     AmbiguousTy(#[label] Span),
+    #[error("unsupported parametric class bound")]
+    #[diagnostic(code("Qsc.TypeCk.UnsupportedParametricClassBound"))]
+    UnsupportedParametricClassBound(#[label] Span),
```
[the remark this answers is not in the document]
Comment (swernli) [reply compiler/qsc_frontend/src/typeck.rs:129]: Wait, never mind. I got confused with the code between the lowerer and the resolver.

File: compiler/qsc_ast/src/ast.rs (hunk 5 above)
Comment (sezna) [reply compiler/qsc_ast/src/ast.rs:1459]: That makes sense, I've gone ahead and removed them. I didn't need them ultimately, but I thought they might be useful...but dead code is dead code!

File: compiler/qsc_ast/src/ast.rs (hunk 6 above)
Comment (sezna) [reply compiler/qsc_ast/src/ast.rs:2057]: Makes sense!

File: compiler/qsc_ast/src/ast.rs (hunk 5 above)
Comment (sezna) [reply compiler/qsc_ast/src/ast.rs:1459]: That makes sense, I've gone ahead and removed them. I didn't need them ultimately, but I thought they might be useful...but dead code is dead code!

File: compiler/qsc_doc_gen/src/display.rs (hunk 12 above)
Comment (sezna) [reply compiler/qsc_doc_gen/src/display.rs:622]: Went ahead and added a printout of bounds here using `ClassConstraints`' `Display` impl.

File: compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs (hunk 13 above)
Comment (sezna) [reply compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs:109]: I think this one is going to be a fast-follow, I'd rather get the feature in for the release and track this one down shortly after. That being said, I'll track this down as soon as I'm done with the other things on this branch...so maybe I'll finish before it goes in. Tl;dr I'd rather this not block the PR, and I'll fix it ASAP, whether that be before or after it goes in.

File: compiler/qsc_frontend/src/typeck.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (sezna) [reply compiler/qsc_frontend/src/typeck.rs:129]: Using `UnrecognizedClass` works for me! Updated.

File: compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs (hunk 13 above)
Comment (swernli) [reply compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs:109]: Sounds good!

**Review by swernli — APPROVED** [review]
Signing off on the compiler changes, thought I see there is some good feedback from Mine still pending. Assuming that goes well, I think this is ready to merge.

File: language_service/src/completion/tests.rs
Comment (sezna) [reply language_service/src/completion/tests.rs:1]: image

File: compiler/qsc_fir/src/ty.rs [hunk 15]
```diff
@@ -178,20 +178,77 @@ fn instantiate_arrow_ty<'a>(
...
+    Iterable {
+        item: Ty,
+    },
+
+    /// A class that is not built-in to the compiler.
+    NonNativeClass(Rc<str>),
+    Num,
+    Integral,
+    Show,
+}
```
Comment (orpuente-MS) [inline compiler/qsc_fir/src/ty.rs:227]: Can we add docstrings to the other variants besides `NonNativeClass`?

File: compiler/qsc_fir/src/ty.rs [hunk 16]
```diff
@@ -178,20 +178,77 @@ fn instantiate_arrow_ty<'a>(
...
+                .collect::<Vec<_>>()
+                .join(", ");
+            write!(f, "{bounds}")
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub enum ClassConstraint {
+    #[default]
```
Comment (orpuente-MS) [inline compiler/qsc_fir/src/ty.rs:212]: Is it correct to say that the default of a `ClassConstraint` is the `Eq` constraint? Where do we use this?

File: compiler/qsc_fir/src/ty.rs [hunk 17]
```diff
@@ -178,20 +178,77 @@ fn instantiate_arrow_ty<'a>(
...
+    Exp {
+        power: Ty,
+    },
+    Iterable {
+        item: Ty,
+    },
+
+    /// A class that is not built-in to the compiler.
+    NonNativeClass(Rc<str>),
+    Num,
```
Comment (orpuente-MS) [inline compiler/qsc_fir/src/ty.rs:224]: As we discussed, it is better if we split the `Num` constraint into `Mul`, `Div`, `Sub`, etc. to stay consistent with `Add` and `Exp` being their own contraints.

File: wasm/src/language_service.rs [hunk 18]
```diff
@@ -164,6 +164,7 @@ impl LanguageService {
                         qsls::protocol::CompletionItemKind::Variable => "variable",
                         qsls::protocol::CompletionItemKind::TypeParameter => "typeParameter",
                         qsls::protocol::CompletionItemKind::Field => "field",
+                        qsls::protocol::CompletionItemKind::Class => "class",
```
Comment (minestarks) [inline wasm/src/language_service.rs:167]: Oh I almost forgot. You need to handle this enum case in both the vscode and playground and map them to the corresponding enum values in the VS Code and Monaco APIs. (There should be two changes, under `vscode/` and `playground/`). That way the proper icon will show up in the UI.

**Review by minestarks — APPROVED** [review]
Looking forward to this going live, thanks!

File: language_service/src/name_locator.rs
Comment (sezna) [reply language_service/src/name_locator.rs:1]: https://github.com/user-attachments/assets/6eb1cd71-8d45-4a6c-b05a-eaf1b31c0c1e

fixed!

File: compiler/qsc_fir/src/ty.rs (hunk 16 above)
Comment (sezna) [reply compiler/qsc_fir/src/ty.rs:212]: This is a bad copy-paste on my part -- we have to have a default in the parser for parser recovery (a pattern I'm not actually very happy with) and I copy/pasted it here.

It hasn't been used anywhere so I removed it. Good catch, thank you.
